pkg/apachepulsar/apachepulsar.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apachepulsar
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+	sourcesdk "github.com/numaproj/numaflow-go/pkg/sourcer"
+)
+
+type PulsarSource struct {
+	client   pulsar.Client
+	consumer pulsar.Consumer
+	toAckSet map[string]pulsar.Message
+	lock     *sync.Mutex
+}
+
+func NewPulsarSource(client pulsar.Client, consumer pulsar.Consumer) *PulsarSource {
+	return &PulsarSource{client: client, consumer: consumer, toAckSet: map[string]pulsar.Message{}, lock: new(sync.Mutex)}
+}
+
+func (ps *PulsarSource) Read(_ context.Context, readRequest sourcesdk.ReadRequest, messageCh chan<- sourcesdk.Message) {
+	ctx, cancel := context.WithTimeout(context.Background(), readRequest.TimeOut())
+	defer cancel()
+	// If we have un-acked data, we return without reading any new data.
+	if len(ps.toAckSet) > 0 {
+		return
+	}
+	for i := 0; i < int(readRequest.Count()); i++ {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			ps.lock.Lock()
+			msg, err := ps.consumer.Receive(ctx)
+			if err != nil {
+				log.Printf("error receiving message %s", err)
+			}
+			messageCh <- sourcesdk.NewMessage(
+				msg.Payload(),
+				sourcesdk.NewOffset([]byte(msg.ID().String()), 0),
+				msg.PublishTime(),
+			)
+			ps.toAckSet[msg.ID().String()] = msg
+			ps.lock.Unlock()
+		}
+	}
+}
+
+func (ps *PulsarSource) Pending(_ context.Context) int64 {
+	return int64(len(ps.toAckSet))
```

Comment:
Do we have a better way to retrieve `pending` other than using toAckSet? `toAckSet` is a workaround when the source client is not able to determine its messages to ack. In sqs source we use it because the client is returning the global msgs to ack. We should avoid using the toAckSet approach when possible.

Since Apache Pulsar is an alternative for Kafka, could you look into the kafka source implementation and see if Apache Plulsar provides similar interfaces to calculate pending?